File: src/Merge Two Sorted Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *mergeTwoLists(ListNode *l1, ListNode *l2) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        if(l1 == 0)
            return l2;
        if(l2 == 0)
            return l1;
        if(l1->val <= l2->val)
            return merge(l1, l2);
        else
            return merge(l2, l1);
    }
    ListNode *merge(ListNode *l1, ListNode *l2)
    {
        ListNode *head = l1;
        ListNode *p, *q, *r, *s;
        do
        {
        
            p = l1;
            q = l1->next;
            r = l2;
            while(l2 && q && l2->val <= q->val)
            {
                s = l2;
                l2 = l2->next;
            }
            if(r != l2)
            {
                p->next = r;
                s->next = q;
            }
            l1 = q;
        } while(l1);
        if(l2)
        {
            p->next = l2;
        }
        return head;
        
    }
    
};

File: python/Merge Two Sorted Lists.py
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
 
class Solution:
    # @param two ListNodes
    # @return a ListNode
    def mergeTwoLists(self, l1, l2):
        flag = True;
        head = None; curr = ListNode(0)
        p1 = l1; p2 = l2
        while p1 and p2:
            if p1.val <= p2.val:
                curr.next = ListNode(p1.val); p1 = p1.next
            else:
                curr.next = ListNode(p2.val); p2 = p2.next
            if flag:
                head = curr.next; flag = False
            curr = curr.next
        while p1:
            curr.next = ListNode(p1.val); p1 = p1.next
            if flag:
                head = curr.next; flag = False
            curr = curr.next
        while p2:
            curr.next = ListNode(p2.val); p2 = p2.next
            if flag:
                head = curr.next; flag = False
            curr = curr.next
        return head